debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/BinlogReader.java

```diff
@@ -410,6 +415,12 @@ protected void handleEvent(Event event) {
 
         // If there is a handler for this event, forward the event to it ...
         try {
+            // Generate heartbeat message if the time is right
+            if (heartbeatTimeout.expired()) {
```

Comment:
This approach won't work if there's no event to handle for a longer time (no changes to the source DB at all). I think we should rather drive this through a separate thread, ideally encapsulated in a separate class which can be re-used across connectors. Roughly like so:

```
class HeartbeatController {
    scheduler = Executors.scheduledManagedExecutorService();
    scheduler.atFixedRate(emitHeartBeat());
}
```

And in the connectors:

```
heartbeatController = new HearbeatController(this:enqueueHeatBeatRecord, 15, Minutes);

handleEvent() {
    writeLock {
         if(somethingToEnqueue) {
             enqueue();
         }
         
         this.latestOffset = currentOffset;
    }
}

emitHearBeat() {
    writeLock {
        eneque(new HeartbeatRecord());        
    }

```

I.e. enqueuing the actual change records and writing/reading the latest offset must happen in a synchronized manner, so to make sure we don't emit any heartbeats with a on outdated offset.

Reply:
I have to disagree here
1. The original issue was problem with high traffic binlog and low traffic topic. So the solution worked well as it moved offsets forward. If there were no events in binlog there was no need to move offsets.
2. I changed the mechanism to real heartbeat so if BinlogReader is running it generates events regardless of binlog traffic. But I avoided using a separata thread as it would unnecessary complicate the code and the less threads the better. As the main loop is now non-blocking we can insert the heartbeat code execution inside the main loop.

Reply:
Yes, thinking about it a bit more this seems like the right thing to do. Esp. as that way it represents livelyness of the binlog reader thread (which is the interesting one) instead of a separate one. We should just clarify in the docs that it's expected that no heartbeat will be emitted a) during snapshotting and b) during binlog reading if no events ocur.

Reply:
In recent version we are emitting heartbeat events regardless of binlog inactivity